src/Parsers/ParserSelectQuery.cpp

```diff
@@ -93,12 +95,22 @@ bool ParserSelectQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
 
     /// SELECT [DISTINCT] [TOP N [WITH TIES]] expr list
     {
+        bool has_all = false;
         if (!s_select.ignore(pos, expected))
             return false;
 
+        if (s_all.ignore(pos, expected))
+            has_all = true;
+
         if (s_distinct.ignore(pos, expected))
             select_query->distinct = true;
 
+        if (!has_all && s_all.ignore(pos, expected))
+            has_all = true;
+
+        if (has_all && select_query->distinct)
+            throw Exception("Can not use DISTINCT alongside ALL", ErrorCodes::LOGICAL_ERROR);
```

Comment:
It should be syntax error, not logical error.

Reply:
Ok.